tests/0151-purge-brokers.c

```diff
@@ -0,0 +1,410 @@
+                                         broker_id);
+                                TEST_ASSERT(rd_kafka_mock_broker_set_down(
+                                                cluster, broker_id) == 0,
+                                            "Failed to set broker %" PRId32
+                                            " down",
+                                            broker_id);
+                                break;
+
+                        case TEST_ACTION_SET_UP_BROKER:
+                                TEST_SAY("Setting up broker %" PRId32 "\n",
+                                         broker_id);
+                                TEST_ASSERT(rd_kafka_mock_broker_set_up(
+                                                cluster, broker_id) == 0,
+                                            "Failed to set broker %" PRId32
+                                            " up",
+                                            broker_id);
+                                break;
+                        case TEST_ACTION_SET_GROUP_COORDINATOR:
+                                TEST_ASSERT(rd_kafka_mock_coordinator_set(
+                                                cluster, "group", TEST_GROUP,
+                                                broker_id) == 0,
+                                            "Failed to set group coordinator "
+                                            "to %" PRId32,
+                                            broker_id);
+                                break;
+                        default:
+                                break;
+                        }
+                }
+
+                fetch_metadata(rk, expected_broker_ids[action],
+                               expected_brokers_cnt[action],
+                               await_after_action_cb, action);
+        }
+        TEST_SAY("Test verification complete\n");
+        rd_atomic32_set(&verification_complete, 1);
+
+        rd_kafka_destroy(rk);
+        test_mock_cluster_destroy(cluster);
+}
+
+/**
+ * @brief Test replacing the brokers in the mock cluster with new ones.
+ *        At each step a majority of brokers are returned by the Metadata call.
+ *        At the end all brokers from the old cluster are removed.
+ */
+static void do_test_replace_with_new_cluster(void) {
+        SUB_TEST_QUICK();
+
+        int32_t expected_brokers_cnt[] = {3, 3, 2, 3, 2, 3, 3, 2, 3};
+
+        int32_t expected_broker_ids[][5] = {{1, 2, 3}, {1, 2, 3}, {2, 3},
+                                            {2, 3, 4}, {3, 4},    {3, 4, 5},
+                                            {3, 4, 5}, {4, 5},    {4, 5, 6}};
+
+        int32_t actions[][2] = {
+            {TEST_ACTION_SET_GROUP_COORDINATOR, 3},
+            {TEST_ACTION_REMOVE_BROKER, 1},
+            {TEST_ACTION_ADD_BROKER, 4},
+            {TEST_ACTION_REMOVE_BROKER, 2},
+            {TEST_ACTION_ADD_BROKER, 5},
+            {TEST_ACTION_SET_GROUP_COORDINATOR, 5},
```

Comment:
How are we verifying that the coordinator is getting updated to the correct one we wanted? This might not be in the scope of this test but then why are we even doing that part?

Reply:
Because if the coordinator doesn't change before removing a broker, that's returned by the `FindCoordinator` request and added again because it's needed as a reference in `rkcg_curr_coord`. So it continues adding that broker again and doesn't reach the desired number of brokers.
In a real cluster this doesn't happen as a new coordinator is elected before the broker shuts down so it's simulating that behaviour.
Initially I wanted to make the coordinator change automatic but that isn't the previous mock cluster contract so I added these manual commands.